streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -998,62 +998,39 @@ trait ZStream[-R, +E, +A] extends Serializable { self =>
-                case Left(Take.Value(a)) =>
-                  f(s, l(a)).flatMap { s =>
-                    if (cont(s)) loop(leftDone, rightDone, s, queue)
-                    else IO.succeed(s)
-                  }
-                case Right(Take.End) =>
-                  if (leftDone) IO.succeed(s)
-                  else loop(leftDone, true, s, queue)
-                case Right(Take.Value(b)) =>
-                  f(s, r(b)).flatMap { s =>
-                    if (cont(s)) loop(leftDone, rightDone, s, queue)
-                    else IO.succeed(s)
-                  }
-              }
-
-          for {
-            queue <- ZManaged.make(Queue.bounded[Elem](capacity))(_.shutdown)
-            _ <- self.fold[R2, E2, A, Unit].flatMap { fold =>
-                  fold((), _ => true, (_, a) => queue.offer(Left(Take.Value(a))).unit)
-                    .foldCauseM(
-                      e => ZManaged.fromEffect(queue.offer(Left(Take.Fail(e)))),
-                      _ => ZManaged.fromEffect(queue.offer(Left(Take.End)))
-                    )
-                    .unit
-                    .fork
-                }
-
-            _ <- that.fold[R2, E2, B, Unit].flatMap { fold =>
-                  fold((), _ => true, (_, a) => queue.offer(Right(Take.Value(a))).unit)
-                    .foldCauseM(
-                      e => ZManaged.fromEffect(queue.offer(Right(Take.Fail(e)))),
-                      _ => ZManaged.fromEffect(queue.offer(Right(Take.End)))
-                    )
-                    .unit
-                    .fork
-                }
+  )(l: A => C, r: B => C): ZStream[R1, E1, C] = {
+
+    def coordinate(ref: Ref[Boolean], dest: Queue[Take[E1, C]]): ZIO[Any, Nothing, Unit] =
+      ref.modify((_, true)).flatMap(p => if (p) dest.offer(Take.End).unit else ZIO.unit)
+
+    val queue = for {
+      ref   <- ZManaged.fromEffect(Ref.make(false))
+      queue <- ZManaged.make(Queue.bounded[Take[E1, C]](capacity))(_.shutdown)
+      _ <- self.fold[R1, E1, A, Unit].flatMap { fold =>
+            fold((), _ => true, (_, a) => queue.offer(Take.Value(l(a))).unit)
+              .foldCauseM(
+                e => ZManaged.fromEffect(queue.offer(Take.Fail(e))),
+                _ => ZManaged.fromEffect(coordinate(ref, queue))
+              )
+              .unit
+              .fork
+          }
+      _ <- that.fold[R1, E1, B, Unit].flatMap { fold =>
+            fold((), _ => true, (_, b) => queue.offer(Take.Value(r(b))).unit)
+              .foldCauseM(
+                e => ZManaged.fromEffect(queue.offer(Take.Fail(e))),
+                _ => ZManaged.fromEffect(coordinate(ref, queue))
+              )
+              .unit
+              .fork
+          }
+    } yield queue
 
-            s <- ZManaged.fromEffect(loop(false, false, s, queue))
-          } yield s
-        }
-    }
+    for {
+      queue  <- ZStream.managed(queue)
+      result <- ZStream.combine(self, that)((_, _) => queue.take)
```

Comment:
😂

I think this could be implemented if `ZStream.combine` allowed the function to keep a state between the invocations, like an unfold. Then you could use `raceWith` to race takes against the two queues. Whenever one side wins, you pass the pending side through the state to the next invocation. That would save the need for the output queue here.

That said, we can defer that to a follow-up ticket that enhances `combine` and implements `mergeWith` in terms of it if you'd like to get interleave in sooner.

Reply:
Yes, I think that makes sense. I guess we could potentially use a `Ref` to pass the loser from one invocation to another like I did to manage state in the other combinators. But I was definitely feeling the lack of ability to maintain some state in writing them. It would probably be both easier to use and more performant to pass the state explicitly.